acme/challenge.go

```diff
@@ -584,37 +223,48 @@ func (dc *dns01Challenge) validate(db nosql.DB, jwk *jose.JSONWebKey, vo validat
 		}
 	}
 	if !found {
-		if err = dc.storeError(db,
-			RejectedIdentifierErr(errors.Errorf("keyAuthorization "+
-				"does not match; expected %s, but got %s", expectedKeyAuth, txtRecords))); err != nil {
-			return nil, err
-		}
-		return dc, nil
+		return storeError(ctx, ch, db, NewError(ErrorRejectedIdentifierType,
+			"keyAuthorization does not match; expected %s, but got %s", expectedKeyAuth, txtRecords))
 	}
 
 	// Update and store the challenge.
-	upd := &dns01Challenge{dc.baseChallenge.clone()}
-	upd.Status = StatusValid
-	upd.Error = nil
-	upd.Validated = time.Now().UTC()
+	ch.Status = StatusValid
+	ch.Error = nil
+	ch.ValidatedAt = clock.Now().Format(time.RFC3339)
 
-	if err := upd.save(db, dc); err != nil {
-		return nil, err
+	if err = db.UpdateChallenge(ctx, ch); err != nil {
+		return WrapErrorISE(err, "error updating challenge")
 	}
-	return upd, nil
+	return nil
 }
 
-// getChallenge retrieves and unmarshals an ACME challenge type from the database.
-func getChallenge(db nosql.DB, id string) (challenge, error) {
-	b, err := db.Get(challengeTable, []byte(id))
-	if nosql.IsErrNotFound(err) {
-		return nil, MalformedErr(errors.Wrapf(err, "challenge %s not found", id))
-	} else if err != nil {
-		return nil, ServerInternalErr(errors.Wrapf(err, "error loading challenge %s", id))
-	}
-	ch, err := unmarshalChallenge(b)
+// KeyAuthorization creates the ACME key authorization value from a token
+// and a jwk.
+func KeyAuthorization(token string, jwk *jose.JSONWebKey) (string, error) {
+	thumbprint, err := jwk.Thumbprint(crypto.SHA256)
 	if err != nil {
-		return nil, err
+		return "", WrapErrorISE(err, "error generating JWK thumbprint")
+	}
+	encPrint := base64.RawURLEncoding.EncodeToString(thumbprint)
+	return fmt.Sprintf("%s.%s", token, encPrint), nil
+}
+
+// storeError the given error to an ACME error and saves using the DB interface.
+func storeError(ctx context.Context, ch *Challenge, db DB, err *Error) error {
+	ch.Error = err
+	if err := db.UpdateChallenge(ctx, ch); err != nil {
```

Comment:
I was trying to ere on the side of being permissive since the client can send multiple retry requests. Instead I'll set to `invalid` if the challenge validation request is successful (the challenge server responds), but the underlying validation does not match.